core/src/main/java/com/linecorp/armeria/common/stream/EventLoopStreamMessage.java

```diff
@@ -0,0 +1,446 @@
+        this.eventLoop = requireNonNull(eventLoop, "eventLoop");
+        queue = new ArrayDeque<>();
+    }
+
+    @Override
+    public boolean isOpen() {
+        return isOpen;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return !isOpen() && !wroteAny;
+    }
+
+    @Override
+    void subscribe(SubscriptionImpl subscription) {
+        Subscriber<?> subscriber = subscription.subscriber();
+        if (!subscribedUpdater.compareAndSet(this, 0, 1)) {
+            eventLoop.execute(() -> failLateSubscriber(this.subscription, subscriber));
+            return;
+        }
+
+        if (eventLoop.inEventLoop()) {
+            doSubscribe(subscription);
+        } else {
+            eventLoop.execute(() -> doSubscribe(subscription));
+        }
+    }
+
+    @Override
+    public void close() {
+        if (eventLoop.inEventLoop()) {
+            doClose(null);
+        } else {
+            eventLoop.execute(() -> doClose(null));
+        }
+    }
+
+    @Override
+    public void close(Throwable cause) {
+        requireNonNull(cause, "cause");
+        if (cause instanceof CancelledSubscriptionException) {
+            throw new IllegalArgumentException("cause: " + cause + " (must use Subscription.cancel())");
+        }
+
+        if (eventLoop.inEventLoop()) {
+            doClose(cause);
+        } else {
+            eventLoop.execute(() -> doClose(cause));
+        }
+    }
+
+    @Override
+    public void abort() {
+        if (abortedUpdater.compareAndSet(this, 0, 1)) {
+            // Let readers of isOpen know immediately that the stream was aborted.
+            isOpen = false;
+
+            if (subscribedUpdater.compareAndSet(this, 0, 1)) {
+                if (eventLoop.inEventLoop()) {
+                    doSetAbortedSubscription();
+                    doCancelOrAbort(false);
+                } else {
+                    eventLoop.execute(() -> {
+                        doSetAbortedSubscription();
+                        doCancelOrAbort(false);
+                    });
+                }
+                return;
+            }
+
+            cancelOrAbort(false);
+        }
+    }
+
+    @Override
+    long demand() {
+        return demand;
+    }
+
+    @Override
+    void request(long n) {
+        if (eventLoop.inEventLoop()) {
+            doRequest(n);
+        } else {
+            eventLoop.execute(() -> doRequest(n));
+        }
+    }
+
+    @Override
+    void cancel() {
+        cancelOrAbort(true);
+    }
+
+    @Override
+    void addObject(T obj) {
+        wroteAny = true;
+        if (eventLoop.inEventLoop()) {
+            doAddObject(obj);
+        } else {
+            eventLoop.execute(() -> doAddObject(obj));
+        }
+    }
+
+    @Override
+    void addObjectOrEvent(Object obj) {
+        if (eventLoop.inEventLoop()) {
+            doAddObjectOrEvent(obj);
+        } else {
+            eventLoop.execute(() -> doAddObjectOrEvent(obj));
```

Comment:
Just curious. Is there any unexpected chance to make added objects mis-ordered? I imagined the scenario as follows:

1. Thread A(not a eventloop) calls addObjectOrEvent with obj 1. -> into the eventloop queue.
2. IO Event occurs on that eventloop. 
3. While the eventloop handles the IO event, addObjectOrEvent is called with obj 2. -> add obj 2 to the queue.
4. After handling IO event, eventloop handles pending tasks. -> add obj 1 to the queue.

I'm not sure if this kind of scenario is possible. Also I guess it's very rare even if possible.

Reply:
I believe you mean obj2 is added from an event loop thread so gets added immediately? It means there is Thread A and Event Loop (e.g., Thread B) writing simultaneously to the stream. I don't think there's any way to guarantee ordering when multiple threads can write simultaneously to a queue. Admittedly, with an event loop, there is a wider window for where such a race could happen, but since any multithreaded writes still have the problem to some degree, I wouldn't worry about it.

Reply:
Ok, it sounds reasonable. Thank you for answer. :-)